libmscore/element.cpp

```diff
@@ -230,6 +230,30 @@ void Element::deleteLater()
 //---------------------------------------------------------
 
 void Element::scanElements(void* data, void (* func)(void*, Element*), bool all)
+{
+    // if leaf node, apply function on this element (after checking if it is visible)
+    // otherwise, recursively apply scanElements to all children
+    // (Note: apply function even if non-leaf node, to the following element types:
+    //        {PAGE, NOTE, REST, GLISSANDO, VBOX, TUPLET} )
```

Comment:
I applying the function to all nodes, surprisingly it seemed to work with no obvious bugs. But I couldn't select measures as I could earlier, by clicking on the staff, the system got selected instead. There might be other subtle bugs like this, but I could only find one right now.

Reply:
Interesting! Normally systems cannot be selected. The same goes for pages, but they have special handling [here](https://github.com/musescore/MuseScore/blob/6d5f118c13d36d21400c625234e7f9e76c25497a/libmscore/mscoreview.cpp#L47-L48). Perhaps if you handle systems in a similar way then the bug will disappear?

Note that the bug has nothing to do with drawing. [Element::draw](https://github.com/musescore/MuseScore/blob/944fe1b40a7eab329a2e6eec2cbdd53a79fb16e3/libmscore/element.h#L319) is a virtual function that does nothing by default, and it is not overridden by [System](https://github.com/musescore/MuseScore/blob/944fe1b40a7eab329a2e6eec2cbdd53a79fb16e3/libmscore/system.h#L79). The bug actually occurs [here](https://github.com/musescore/MuseScore/blob/9fb78e4ca7ed1327544debd3808260facd19e568/libmscore/page.cpp#L274), where `scanElements` is used to build a [BSP Tree](https://github.com/musescore/MuseScore/blob/c9f8a138529c15bf9d096e53a4a2af14af32ae11/libmscore/bsp.h#L32) of elements that will be searched to find the element at the given position. System is added to that tree when it wasn't previously.

If bugs occur with other classes then we should aim to fix them in the same way as the drawing case: Rather than skipping elements, `scanElements` should visit every element, but the effect of calling `func` on certain elements should be no-op.